Legacy data result pruning and basic pruning

hacheigriega opened the pull request:
## Motivation

Batching pruning addresses following issues:

1. We need to prune the accumulated batches and their associated data in batches.
2. Since legacy data results cannot be retrieved based on a batch number, they have to be pruned separately.
3. We need a permanent, efficient solution for batch pruning.

Issue 1 was implemented in #664. This PR addresses the other two issues.


## Explanation of Changes

### Pruning Parameters

Pruning logic is parameterized by the following module states, which will be added by the upgrade handler:

- `MaxBatchPrunePerBlock` is a module parameter that specifies the maximum number of batches to prune per block in the Batch Pruning strategy. (Defaults to 100)
- `MaxLegacyDataResultPrunePerBlock` is a module parameter that specifies the maximum number of legacy data results to prune per block in the Legacy Data Result Pruning strategy. (Defaults to 1000)
- `BatchNumberAtUpgrade` is the batch number of the latest batch at the time of upgrade. The upgrade handler should set the value in order to record the batch number of the last batch containing legacy data results.
- `HasPruningCaughtUp` is set to true when either of the following conditions is met:
    - (i) All batches up to batchNumberAtUpgrade have been pruned.
    - (ii) All batches up to `currentBatchNum - NumBatchesToKeep` have been pruned.
    
    It is set to false at the time of upgrade, initiating the Batch Pruning strategy. Once it is switched to true, the batching end blocker terminates the Batch Pruning strategy and initiates the Legacy Data Result Pruning strategy.

There is also a new global variable: 

- `NumBatchesToKeep` specifies the number of batches to preserve in the state. Currently set to 10000.
    - This value is not kept as a module state because it is a fixed value that originates from outside the module logic. In a future upgrade we may turn this into a module parameter, in which case the pruning logic would need to be updated to accommodate its variability.

### Pruning Strategies

**Basic Pruning (Implemented in this PR)**
When a new batch is created, the Basic Pruning strategy prunes one batch at `NewBatchNumber - NumBatchesToKeep`, if this number is higher than `BatchNumberAtUpgrade`.

We also temporarily add Batch Pruning and Legacy Data Result Pruning strategies to prune the batches and their data that have been accumulated until the upgrade.

**Batch Pruning (Implemented in #664)**
This pruning strategy prunes up to MaxBatchPrunePerBlock batches and their associated data, except for data results and their batch assignments, every block either up to batch at BatchNumAtUpgrade or up to batch at currentBatchNum - NumBatchesToKeep. Once either of the two termination conditions is met, HasPruningCaughtUp is switched to true, terminating Batch Pruning and initiating Legacy Data Result Pruning.

**Legacy Data Result Pruning (Implemented in this PR)**
Once `HasPruningCaughtUp` is switched to true, this pruning strategy begins. It iterates through legacy data result collection, pruning at most  `MaxLegacyDataResultPrunePerBlock` data results and their batch assignments per block.


## Related PRs and Issues

Closes #671

hacheigriega committed:
feat(x/batching): batch prune batches and data results

Prune batches and their associated data at every block based on two new
module parameters NumBatchesToKeep and MaxBatchPrunePerBlock.
For pruning data results and their batch assignment data, we resort to
naive implementation because there is no mapping to data result objects
from batch number or data result ID. In this implementation we go through
`MaxDataResultsToCheckForPrune` items in the store starting from a random
point and delete those whose associated batches have been pruned.

hacheigriega committed:
feat(x/batching): implementation of batching module pruning

Implementation of batching module pruning as described in #673.

hacheigriega committed:
refactor(x/batching): simplify batch pruning

Simplify batch pruning logic by fixing numBatchesToKeep.

hacheigriega committed:
chore(x/batching): set version to 2 and add migrator from version 1

hacheigriega committed:
fix(x/batching): proper registration of MsgUpdateParams

Thomasvdam commented on x/batching/keeper/batch_assignments.go:
```diff
@@ -0,0 +1,67 @@
+	return batchAssignments, err
+}
+
+func (k Keeper) SetBatchDataResults(ctx context.Context, batchNumber uint64, dataRequestIDHeights types.DataRequestIDHeights) error {
```
Is this method used?

hacheigriega replied:
Explains the issue with data result pruning after upgrade 😅 Thank you.

Thomasvdam commented on x/batching/keeper/pruning.go:
```diff
@@ -0,0 +1,227 @@
+	dataResults, err := k.GetBatchDataResults(ctx, batchNumToPrune)
+	if err != nil {
+		if errors.Is(err, collections.ErrNotFound) {
+			k.Logger(ctx).Info("cannot prune batch because schema change has not been applied", "batch_num", batchNumToPrune)
```
This is not always true right? We can also have batches that only contain validator changes and no results.

hacheigriega replied:
Good catch.. We store an empty slice when constructing a batch so this does not give an error.

Thomasvdam reviewed (commented):
Overall I think this looks good. I'm not sure if we did a double check of the proposed strategies and numbers with Mario.

hacheigriega committed:
refactor(x/batching): update in hasPruningCaughtUp logic

Before this commit, hasPruningCaughtUp was switched to true only when
BatchPruneBatches has pruned all batched up to batchNumberAtUpgrade.
This condition could be blocked for a long time by a big numBatchToKeep.
So we add an alternative condition under which hasCaughtUp is switched to
true: If all batches up to (currentBatchNum - numBatchesToKeep) have been
pruned.
BatchPruneBatches is updated accordingly. BasicPruneBatch now prunes a
batch as long as it exists, whether it was created before or after the
upgrade.

hacheigriega committed:
fix(x/batching): create map from batch number to data results

hacheigriega committed:
chore: address lint warnings
